Sources/_StringProcessing/ByteCodeGen.swift

```diff
@@ -35,6 +37,12 @@ extension Compiler.ByteCodeGen {
     case let .backreference(ref):
       try emitBackreference(ref)
 
+    case let .symbolicReference(id):
+      guard let offset = referencedCaptureOffsets[id] else {
+        throw RegexCompilationError.uncapturedReference
```

Comment:
I don't think this should be an error — a valid regex can include a back reference before its been matched. 

Consider `(?:\w\1|:(\w):)+` — the first time through the outer group, `\1` doesn't have a value, so the first branch of the alternation can't match. Once the second branch matches something, however, `\1` _does_ have a value, and participates in matching in the way you would expect. So that pattern matches `:a:bacada`, but not `bacada` on its own. (https://regex101.com/r/mMxqcc/1)

Reply:
Updated.

Reply:
Update to support post-hoc captures. However, I think we should still emit an error when a reference is not captured anywhere in the regex, as it seems consistent with the error here https://regex101.com/r/vvXsgy/1.

Reply:
@hamishknight how does the parser handle this? Maybe it's the same kind of strategy.

Reply:
The parser doesn't currently diagnose this at all, it'll be a part of the post-processing pass over the AST which we have yet to implement (https://github.com/apple/swift-experimental-string-processing/issues/116). I think we'll want the same diagnostic behavior as proposed here, i.e allow forward references, but error on references to captures that don't exist anywhere in the pattern.